marker_rustc_driver/src/main.rs

```diff
@@ -71,6 +95,49 @@ impl rustc_driver::Callbacks for MarkerCallback {
     }
 }
 
+fn register_tracked_env(sess: &mut rustc_session::parse::ParseSess, vars: &[(&'static str, String)]) {
+    use rustc_span::Symbol;
+    let env = sess.env_depinfo.get_mut();
+
+    for (key, value) in vars {
+        env.insert((Symbol::intern(key), Some(Symbol::intern(value))));
+    }
+}
+
+fn register_tracked_files(sess: &mut rustc_session::parse::ParseSess, lint_crates: &[LintCrateInfo]) {
+    use rustc_span::Symbol;
+
+    let files = sess.file_depinfo.get_mut();
+
+    // Cargo sets the current directory, to the folder containing the `Cargo.toml`
+    // file of the compiled crate. Therefore, we can use the relative path.
+    let cargo_file = Path::new("./Cargo.toml");
+    if cargo_file.exists() {
+        files.insert(Symbol::intern("./Cargo.toml"));
+    }
+
+    // At this point, the lint crate paths should be absolute. Therefore, we
+    // can just throw it in the `files` set.
+    for lint_crate in lint_crates {
+        if let Some(name) = lint_crate.path.to_str() {
+            files.insert(Symbol::intern(name));
+        } else {
+            // The name is not a valid UTF-8 String. This is not ideal but at the
+            // same time not problematic enough to throw an error. It just means
+            // that the lint crate can't be tracked and the driver might reuse
+            // a cache if only that crate was changed.
```

Comment:
I recommend you not to care about supportin non-utf8 paths. There are a bunch of reasons to do so. First of all, it simplifies the code, and second a lot of tools assume they work with UTF8 paths, for example cargo-metadata crate does so. You may see that it uses [`Utf8Path[Buf]`](https://docs.rs/cargo_metadata/latest/cargo_metadata/struct.Metadata.html#structfield.target_directory) for storing all the paths parsed from `cargo metadata`.

My first impression when I saw this in `cargo-metadata` was "why do you limit the users to UTF8", but then I read the reasoning from [`camino`](https://github.com/camino-rs/camino) crate that implements the `Utf8Path[Buf]`, and was convinced that this is the right thing to do. I never in my life had to work with non-UTF8 paths. If you ever happen to work with something like that you are already fucked, and cargo-marker won't be the only tool that doesn't work for you. Can you imagine someone invoking `cargo-marker` on a project that uses invalid UTF8 in its code structure? I can't, and thus I propose adopting `camino` and releaving this mental complexity from the codebase. My question is wether you would accept a PR with such kind of change?

Reply:
I second the notion here, if cargo metadata already assumes utf-8 paths, we should lean on that for simplicity. I think this is the right design decision.

Reply:
I'm fine with moving to UTF-8 based paths. If we document it, it should be a fine restriction to make. Though I would not do this in this PR